pkg/hubble/exporter/aggregator.go

```diff
@@ -0,0 +1,187 @@
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	fa "github.com/cilium/cilium/pkg/hubble/parser/fieldaggregate"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// AggregateKey represents a combination of field values used for aggregation.
+type AggregateKey string
+
+// AggregateValue holds the counters for ingress, egress, and unknown direction flows.
+type AggregateValue struct {
+	IngressFlowCount          int
+	EgressFlowCount           int
+	UnknownDirectionFlowCount int
+	ProcessedFlow             *flowpb.Flow
+}
+
+type Aggregator struct {
+	m              map[AggregateKey]*AggregateValue
+	mu             lock.RWMutex
+	fieldAggregate fa.FieldAggregate
+	logger         *slog.Logger
+}
+
+// NewAggregator creates a new aggregator without field aggregation
+func NewAggregator(logger *slog.Logger) *Aggregator {
+	return &Aggregator{
+		m:      make(map[AggregateKey]*AggregateValue),
+		logger: logger.With(logfields.LogSubsys, "hubble-aggregator"),
+	}
+}
+
+// NewAggregatorWithFields creates a new aggregator with field aggregation
+func NewAggregatorWithFields(fieldAggregate fa.FieldAggregate, logger *slog.Logger) *Aggregator {
+	return &Aggregator{
+		m:              make(map[AggregateKey]*AggregateValue),
+		fieldAggregate: fieldAggregate,
+		logger:         logger.With(logfields.LogSubsys, "hubble-aggregator"),
+	}
+}
+
+func (a *Aggregator) Add(ev *v1.Event) {
+	f := ev.GetFlow()
+	if f == nil {
+		return
+	}
+
+	processedFlow := &flowpb.Flow{}
+	a.fieldAggregate.Copy(processedFlow.ProtoReflect(), f.ProtoReflect())
+
+	k := a.generateAggregationKey(processedFlow)
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	v, ok := a.m[k]
+	if !ok {
+		switch f.GetTrafficDirection() {
+		case flowpb.TrafficDirection_INGRESS:
+			v = &AggregateValue{
+				IngressFlowCount: 1,
+				ProcessedFlow:    processedFlow,
+			}
+		case flowpb.TrafficDirection_EGRESS:
+			v = &AggregateValue{
+				EgressFlowCount: 1,
+				ProcessedFlow:   processedFlow,
+			}
+		default:
+			v = &AggregateValue{
+				UnknownDirectionFlowCount: 1,
+				ProcessedFlow:             processedFlow,
+			}
+		}
+		a.m[k] = v
+	} else {
+		switch f.GetTrafficDirection() {
+		case flowpb.TrafficDirection_INGRESS:
+			v.IngressFlowCount++
+		case flowpb.TrafficDirection_EGRESS:
+			v.EgressFlowCount++
+		default:
+			v.UnknownDirectionFlowCount++
+		}
+	}
+}
+
+func (a *Aggregator) generateAggregationKey(processedFlow *flowpb.Flow) AggregateKey {
+	b, _ := proto.Marshal(processedFlow.ProtoReflect().Interface())
+	return AggregateKey(b)
+}
+
+// AggregatorRunner manages the lifecycle of an aggregator with periodic export.
+type AggregatorRunner struct {
+	aggregator *Aggregator
+	interval   time.Duration
+	encoder    Encoder
+	logger     *slog.Logger
+
+	stop chan struct{}
+	wg   sync.WaitGroup
+}
+
+func (r *AggregatorRunner) Start() {
+	r.stop = make(chan struct{})
+	r.wg.Add(1)
+	go r.run()
+}
```

Comment:
suggestion/correctness/nit: If `Start()` is called twice before `Stop()`, `r.stop` will get replaced and `Stop()` won't be able to reach the original value. This also means the first run goroutine will leak and never stop. Since it never stops, the WaitGroup is never decremented and calls to `Stop()` would block forever. We should probably add a check in `Start` that if `r.stop` is not nil, panic.

I mark this as nit because the way we use the AggregatorRunner right now, it can't be called twice, so there's not immediate danger.